magento/pwa-studio
# Refactor <img /> tags to utilize srcset and sizes for optimized images.

## Description

As of today, we don't utilize `srcset` and `sizes` for `img` tags due to which, a high-resolution image will be downloaded where a medium one would suffice. By giving `sizes` attribute, we will be letting the browser know about the size of the image that should be rendered and based on `window.devicePixelRatio` the browser will what image to use from the `srcset`.

## Related Issue
Closes #1320.

## Verification Steps
1. Go to any page that has an image.
2. Open elements tab in dev tools and hover over the element that represents the image.
3. Note down the width of the image that is displayed once you hover over it and multiply that with the result of running `window.devicePixelRatio`. Check in the `srcset` of the image element for the option that is closest the number you got after the above multiplication.
4. Go to the network tab and check if that is the image that is downloaded.
5. Alternatively, once you hover over the image element, it should tell you the `currentSrc` that is used to render that image and in the URL you should see the optimized width and heigh used.

On a simpler note, perform these commands in the console to verify if the right image is downloaded:
1. Hover over the image element that you want to test.
2. Go to the console and find the srcset of the image by typing `const imgSrcset = $0.srcset.split(',')`.
3. Then find the effective resolution of the image that needs to be downloaded by typing `const effectiveWidth = $0.width * window.devicePixelRatio`.
4. Use the value you got in step 3 and find the URL that is closest to that in the array you got in step 2.
5. Now find the src that is used to download the image by typing `const currentSrc = $0.currentSrc`.
6. The result of step 4 and step 5 should be the same.

## Screenshots
![image](https://user-images.githubusercontent.com/35203638/63383516-91b0b880-c362-11e9-9070-f9b46aaaea89.png)


## Checklist
* Functionality should not have changed. This an optimization request, not a feature request.
* Verify against different resolutions and different device types because each has its own pixel densities. For instance, a Galaxy S5 has a pixel density of 3 whereas an iPad has 2.
